### PR TITLE
contracts: migrate contracts and tests from upkeep-contracts repo

### PR DESCRIPTION
**Pull request by Nic0S**
Also necessary helpers to shared testing library and deduplicate
functionality present in both repos.

### CODE REVIEW

File: contracts/package.json [hunk 1]
```diff
@@ -44,5 +44,8 @@
     "ts-node": "^10.0.0",
     "typechain": "^5.0.0",
     "typescript": "^4.3.2"
+  },
+  "dependencies": {
+    "@openzeppelin/contracts": "^4.3.3"
```
Comment (RyanRHall) [inline contracts/package.json:49]: can we move this to `devDependencies` with the others?

File: contracts/test/v0.8/CronUpkeep.test.ts [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    basicSpec = await cron.cronStringToEncodedSpec('0 * * * *')
+  })
+
+  afterEach(async () => {
+    await h.reset()
+  })
+
+  it('has a limited public ABI', () => {
+    // Casting cron is necessary due to a tricky yarn version mismatch issue, likely between ethers
+    // and typechain. Remove once the version issue is resolved.
```
Comment (RyanRHall) [inline contracts/test/v0.8/CronUpkeep.test.ts:111]: can we make a new "bug" ticket for this and add the link here?

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: contracts/test/v0.8/CronUpkeep.test.ts (hunk 2 above)
Comment (Nic0S) [reply contracts/test/v0.8/CronUpkeep.test.ts:111]: done!

File: contracts/package.json (hunk 1 above)
Comment (Nic0S) [reply contracts/package.json]: done!

**Comment by Nic0S** [comment]
I noticed I got a lot of yarn lockfile changes on yarn install, when I would only expect the change for openzeppelin/contract..the same thing is happening for me on the develop branch though.

**Comment by jkongie** [comment]
> I noticed I got a lot of yarn lockfile changes on yarn install, when I would only expect the change for openzeppelin/contract..the same thing is happening for me on the develop branch head though

Probably because I merged a change with a number of new packages to support GraphQL in the operator ui

File: contracts/test/v0.8/Cron.test.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import { validCrons, invalidCrons } from '../test-helpers/fixtures'
+import { reset, setTimestamp } from '../test-helpers/helpers'
+
+let cron: CronInternalTestHelper | CronExternalTestHelper
+let cronInternal: CronInternalTestHelper
+let cronExternal: CronExternalTestHelper
+
+const timeStamp = 32503680000 // Jan 1, 3000 12:00AM
+
+describe('Cron [ @skip-coverage ]', () => {
```
Comment (RyanRHall) [inline contracts/test/v0.8/Cron.test.ts:15]: can we move this to just the individual test that was overloading the coverage analysis?

File: contracts/test/v0.8/CronUpkeep.test.ts [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+
+async function createBasicCron() {
+  return await cron.createCronJobFromEncodedSpec(
+    cronReceiver1.address,
+    handler1Sig,
+    basicSpec,
+  )
+}
+
+describe('CronUpkeep [ @skip-coverage ]', () => {
```
Comment (RyanRHall) [inline contracts/test/v0.8/CronUpkeep.test.ts:63]: same thing, does this need to be on the whole suite?

File: contracts/test/v0.8/CronUpkeep.test.ts (hunk 2 above)
Comment (RyanRHall) [reply contracts/test/v0.8/CronUpkeep.test.ts:111]: nitpick, but it's nice to add a "TODO" at the beginning of the comment so devs know to try to fix this in the future

File: contracts/test/v0.8/CronUpkeepFactory.test.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import type { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
+import { reset } from '../test-helpers/helpers'
+
+let cronExternalLib: Contract
+let factory: CronUpkeepFactory
+
+let admin: SignerWithAddress
+let owner: SignerWithAddress
+
+describe('CronUpkeepFactory [ @skip-coverage ]', () => {
```
Comment (RyanRHall) [inline contracts/test/v0.8/CronUpkeepFactory.test.ts:14]: same thing

File: contracts/test/v0.8/EthBalanceMonitor.test.ts [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+
+let bm: EthBalanceMonitorExposed
+let receiveReverter: ReceiveReverter
+let receiveEmitter: ReceiveEmitter
+let receiveFallbackEmitter: ReceiveFallbackEmitter
+let owner: SignerWithAddress
+let stranger: SignerWithAddress
+let keeperRegistry: SignerWithAddress
+
+describe('EthBalanceMonitor [ @skip-coverage ]', () => {
```
Comment (RyanRHall) [inline contracts/test/v0.8/EthBalanceMonitor.test.ts:56]: same thing

File: tools/ci/solidity_coverage [hunk 7]
```diff
@@ -3,4 +3,4 @@
 set -e
 
 yarn workspace @chainlink/contracts compile
-yarn workspace @chainlink/contracts coverage
+NODE_OPTIONS="--max-old-space-size=5120" yarn workspace @chainlink/contracts coverage
```
Comment (RyanRHall) [inline tools/ci/solidity_coverage:6]: was this only needed for the coverage analysic? or did we have the error on tests too?

**Review by RyanRHall — CHANGES_REQUESTED** [review]

**Comment by RyanRHall** [comment]
> I noticed I got a lot of yarn lockfile changes on yarn install, when I would only expect the change for openzeppelin/contract..the same thing is happening for me on the develop branch head though

Not uncommon for yarn.lock to have lots of changes after changing one dep, because yarn.lock includes all the sub-dependencies. So we added OZ/contracts, but that comes with a bunch of sub-dependencies that need to be version controlled too.

File: tools/ci/solidity_coverage (hunk 7 above)
Comment (RyanRHall) [reply tools/ci/solidity_coverage:6]: Also - can we move this env var to the workflow? That feels like a better place for it than in this script. `.github/workflows/solidity.yml`

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: tools/ci/solidity_coverage (hunk 7 above)
Comment (Nic0S) [reply tools/ci/solidity_coverage:6]: Only for coverage, I haven't seen any failures on the test side

File: .github/workflows/solidity.yml [hunk 8]
```diff
@@ -35,6 +35,8 @@ jobs:
             ${{ runner.os }}-
       - run: yarn install --frozen-lockfile
       - name: Run coverage
+        env:
+          NODE_OPTIONS: "--max-old-space-size=5120"
```
Comment (RyanRHall) [inline .github/workflows/solidity.yml:39]: :+1:

**Review by RyanRHall — APPROVED** [review]

**Pull request merged**